pymc/tests/test_sampling.py

```diff
@@ -1446,3 +1449,171 @@ def test_no_init_nuts_compound(caplog):
         b = pm.Poisson("b", 1)
         pm.sample(10, tune=10)
         assert "Initializing NUTS" not in caplog.text
+
+
+class TestCompileForwardSampler:
+    @staticmethod
+    def get_function_roots(function):
+        roots = []
+
+        def expand(node):
+            if (
+                isinstance(node, Variable)
+                and node.owner is None
+                and getattr(node, "name", None) is not None
+            ):
+                roots.append(node)
+            return node.get_parents()
+
+        list(walk(function.maker.fgraph.outputs, expand))
+        return roots
```

Comment:
Does something like this work?

```suggestion
    return [
        var 
        for var in aesara.graph.basic.graph_inputs(f.maker.fgraph.outputs))
        if var.name 
]
```